pkg/executor/service.go

```diff
@@ -42,31 +43,62 @@ func (sl ServiceLauncher) Launch() (TaskHandle, error) {
 		return nil, err
 	}
 
-	return &ServiceHandle{th}, nil
+	return NewServiceHandle(th), nil
 }
 
-// ServiceHandle is a decorator and TaskHandle implementation that should be used with tasks that do not stop on their own.
-type ServiceHandle struct {
+type serviceHandle struct {
 	TaskHandle
+	taskCanBeTerminated bool
+	err                 error
+	mutex               *sync.Mutex
+}
+
+// NewServiceHandle wraps TaskHandle with serviceHandle.
+func NewServiceHandle(handle TaskHandle) TaskHandle {
+	return &serviceHandle{
+		TaskHandle: handle,
+		mutex:      &sync.Mutex{}}
 }
 
 // Stop implements TaskHandle interface.
-func (s ServiceHandle) Stop() error {
-	if s.TaskHandle.Status() != RUNNING {
-		logrus.Errorf("Stop(): ServiceHandle with command %q has terminated prematurely", s.TaskHandle.Name())
-		logOutput(s.TaskHandle)
-		return errors.Errorf("Service %q has ended prematurely", s.TaskHandle.Name())
+func (s *serviceHandle) Stop() error {
+	err := s.checkError()
+	if err != nil {
+		return err
 	}
 
 	return s.TaskHandle.Stop()
 }
 
 // Wait implements TaskHandle interface.
-func (s ServiceHandle) Wait(duration time.Duration) bool {
-	if s.TaskHandle.Status() != RUNNING {
-		logrus.Errorf("Wait(): ServiceHandle with command %q has terminated prematurely", s.TaskHandle.Name())
-		logOutput(s.TaskHandle)
+func (s *serviceHandle) Wait(duration time.Duration) bool {
+	err := s.checkError()
+	if err != nil {
+		// TODO(skonefal): Return error here when wait will return errors.
+		return s.TaskHandle.Wait(duration)
 	}
 
 	return s.TaskHandle.Wait(duration)
 }
+
+func (s *serviceHandle) checkError() error {
+	s.mutex.Lock()
+	defer s.mutex.Unlock()
+
+	if s.err != nil {
+		return s.err
+	}
+
+	if !s.taskCanBeTerminated {
+		s.taskCanBeTerminated = true
+		if s.TaskHandle.Status() == TERMINATED {
```

Comment:
@skonefal - consider returning an error if status is not equal to `RUNNING`. It would be more future-proof.